facility locator local store rebase

patrickvinograd opened the pull request:
Rebased from master, replaces #567.

saneshark commented:
@patrickvinograd Did you want me to start reviewing this? Should we close #567?

patrickvinograd commented:
Sure - I believe i addressed all the outstanding concerns you and @aub had so another run through would be great.

saneshark commented:
@patrickvinograd looks like there are some conflicts to the above files, but I'll begin reviewing now.

saneshark commented on app/models/va_facility.rb:
```diff
@@ -49,29 +46,25 @@ def self.find_by(id:)
-      CEMETERY => NCAFacilityAdapter.new,
-      BENEFITS => VBAFacilityAdapter.new
+      HEALTH => Facilities::LocalStore.new(ENV['VHA_MAPSERVER_URL'], nil, VHAFacilityAdapter),
+      CEMETERY => Facilities::LocalStore.new(ENV['NCA_MAPSERVER_URL'], nil, NCAFacilityAdapter),
```
This might be for a future refactor, but is it really necessary to have adapters? Why not just 3 separate clients? And a single client that manages the mutex synchronization?

For keeping code DRY you could use mixins for any components that are common to all three clients. I think this would be more the Ruby way.

I think this would help a little bit with readability, knowing why there is a need for 3 separate APIs, etc.

patrickvinograd replied:
I'll consider it. I felt there was testability benefit in keeping the json -> model conversion logic in a simple adapter class that had no request logic, no need for stubbing, etc., so I could blast out a bunch of data format/conversion tests for the noisy data we are getting for these facilities.

saneshark replied:
Chances are only one or more services might be able to be cached so was thinking having more separation of concerns might be more readable, just cache the method that fetches based on a uri for the key or something only if successful.

In which case the separation into distinct classes might make it more readable which of the clients need this. You could also then have custom middleware for each, because it seems as though maybe some of the clients handle errors one way and others handle them another, etc. It would be more composable and readable I think... But again, it's not even a minor gripe. 

The mutexes still seem weird to me, so I'm wondering if maybe having the 1 client to rule 3 others would kind of speak to the need of using mutex better too. But again, just a nitpick here. Something to think about in future refactor maybe.

saneshark commented on lib/facilities/bulk_client.rb:
```diff
@@ -0,0 +1,91 @@
+    rescue Facilities::Errors::ServiceError => e
+      Rails.logger.warn "Facility lastEditDate check request failed: #{e.message}"
+      return nil
+    end
```
Do you really need all these Rails.logger calls? If a Faraday exception is raised isn't whatever is provided in Sentry sufficient to debug the issue?

patrickvinograd replied:
I'm still not 100% clear on what shows up in Sentry vs. not. This exception is rescued and not re-raised, so my expectation was it would not show up in Sentry.

A failure to check the timestamp of the GIS data is not fatal if we already have a cached copy, but it's still something notable that I might want to scan logs for to assess how frequently it's failing.

saneshark replied:
I think I tagged the wrong line. My comment makes no sense for this one, you're right. You're not reraising an exception here.

If it's notable, it might be useful to use statsd maybe instead? it sounds like you intend to use it as more of a performance metric than anything meant for debugging purposes maybe

saneshark commented on lib/facilities/bulk_client.rb:
```diff
@@ -0,0 +1,91 @@
+    rescue Facilities::Errors::ServiceError => e
+      Rails.logger.error "Facility fetch request failed: #{e.message}"
+      raise
+    end
```
Ideally these rescue blocks would be handled in some sort of a "perform" block, all in one place rather than duplicated many times over. I realize that the messages are distinct, but I wonder if that's necessary if you have a stacktrace in Sentry for debugging purposes.

patrickvinograd replied:
I'll look into that.

saneshark commented on lib/facilities/local_store.rb:
```diff
@@ -0,0 +1,88 @@
+      end
+    end
+
+    def check_freshness_and_update
```
How often do facilities and/or locations change? Would it be worth while caching / storing these in Redis so that we don't need to make requests for them, perhaps on a 24 hours window of time?

patrickvinograd replied:
Right now they are changing ~ weekly as we are introducing new aspects of facilities and the administrations are actively cleaning their data.  Longer term it's probably going to be more like monthly or quarterly as administrations do a refresh of their data and it gets pushed to GIS.

I considered and started implementing a Redis approach. I discarded it, mainly for the same reason that Carl's nginx stale caching idea didn't seem right - the typical use is not getting individual resources by ID, it's doing a bounding box query. So even if the full records are stored in redis, I have to store the coordinates and IDs locally, and now I have two levels of caching which seemed over-engineered for this application. 

To be clear, we are checking the GIS data timestamp once an hour, and only pulling down the full dataset if the timestamp indicates that there is new data.

patrickvinograd commented:
Conflicts are just due to the data field changes I had to do yesterday, and don't change the implementation substantively.

saneshark commented:
Overall my comments are all nitpicks and could be addressed in future PRs. If there is a way to cleanup some of the raising and logging stuff that would be awesome. Otherwise, LGTM.

patrickvinograd commented:
I incorporated the review feedback and rebased yet again in branch `pv-fl-local-cache`. However this work is on hiatus pending overall decisions about geocoding accuracy/etc. in the data we are consuming. Closing this PR for now, will update corresponding tickets as well.
